### DISCUSSION TITLE
Improved Directory Structure

### DISCUSSION THREAD

**Discussion by atinux** [nuxt/nuxt#20251] [RFCs]
Nuxt 3 is now a fullstack framework thanks to [Nitro](https://nitro.unjs.io) with the ability to have [auto-imported modules](https://nuxt.com/docs/guide/directory-structure/modules) and [layers](https://nuxt.com/docs/getting-started/layers), I believe it's time to rethink the directory structure to give more separation for the "app" level.

This is the proposal I am thinking of:

```bash
.output/
.nuxt/
app/
  assets/
  components/
  composables/
  layouts/
  middleware/
  pages/
  plugins/
  utils/
  main.vue
  router.options.ts
content/
layers/
modules/
node_modules/
public/
server/
  api/
  middleware
  routes/
  utils/
nuxt.config.ts
```

Do you see any downside of this approach?

The minimal starter would then be:

```bash
app/
  main.vue
public/
  favicon.ico
nuxt.config.ts
```

**Comment by pi0** [comment]
I like the `app` separation makes things much cleaner especially considering nuxt is a full-stack framework and also can be used in contexts like docs that mainly need a top level `content` for example.

This change is kinda mainly defaulting `srcDir` to `app` right? (not exactly but almost)

If we can make changes that `srcDir` defaulting is default, we could easily change starter template to use it as default for new structure what do you think?

**Reply by atinux** [reply]
> This change is kinda mainly defaulting srcDir to app right? (not exactly but almost)

Not that easy from what I tested since `public/` and `content/` are based on `srcDir` too, we may want to update it in many places.

This will be a breaking change though and should be considered for Nuxt 4.

**Reply by pi0** [reply]
That's what i am suggesting. For both we could test if dir exists in `rootDir` and add to the array. At least for `public` it is easy we could add both even nitro ignores the non existing one. Not sure about best way for `content`. Once we overcome this, we could do the change. We _could_ adopt earlier for Nuxt 3 too if solve such issues and make default breaking change in Nuxt 4...

**Comment by flapili** [comment]
Hi, I have few idea/interrogation about DX :

1 ) why not consider the app as a special layer too ? like the error was a special layout in vue 2 ? like `layers/app/` as example .

2 ) in my opinion everything related to the app should be in the app folder, like `public` as example (what if in the future we are able to inherit from layer's public folder too ? but that could lead to overwriting/ name collision (a prefix could be used ?)

3 ) some as point 2, what about server dir ? I didn't think about use cases of inheritance of this dir but there is probably some

4) the nuxtconfig shouldn't be in the app folder too as it's related to the app ?

**Reply by atinux** [reply]
Actually the layers can overwrite most of the directory structure **including server/**. We can potentially extend `public/`, my only concerns was regarding of removing files from a `public/` layer.

**Reply by flapili** [reply]
In my opinion dirs that could be extended should be in the app folder too, else I think some people will think only what is inside /app could be extended 😅

**Comment by cannap** [comment]
What about using a 'config' folder to separate Nuxt configuration files for individual modules or for the entire application?

For example, you could have ` 'config/i18n.ts'`  and `'config/n.app.ts'` for Nuxt-specific configurations. 
To differentiate between module-specific configs, you could use a prefix/suffix such as ` 'm.i18n.ts'`

I know you can also split it yourself

**Reply by atinux** [reply]
Having a `config/` dir could make sense, but Nuxt will have to pull everything together in order to get the final config file.

**Comment by stafyniaksacha** [comment]
I like it, it would allow having more predictable paths, so improving scanning/fs perf

- Is `app/main.vue` the old `app.vue` ? Why not `app/index.vue` (`app/error.vue`)
- How about `*.md` & assets in `pages/`?
- How about autoloading `layers/*` like modules
- Why `app` and not `client`. We could make an app with only server content
- I like the `config/` from @cannap, but may it conflict between layers & module config?

**Reply by atinux** [reply]
> Is app/main.vue the old app.vue ? Why not app/index.vue (app/error.vue)

Yes, I thought about `app/index.vue` but feels confusing, and `main.vue` is more close to `main.js` like in normal Vue projects

> How about *.md & assets in pages/?

Not sure but I don't see an issue of having `app/pages/*md` or maybe thinking of having `content/pages/*.md` 🤔 

> How about autoloading layers/* like modules

I thought about it as well 😄 

> Why app and not client. We could make an app with only server content

Because app is also having SSR and client will be confusing, also we have `.server.` suffix for server components

> I like the config/ from @cannap, but may it conflict between layers & module config?

I like it too and we had a RFC for it if I remember well. I also need to think about where the `app.config.ts` lives with this new structure, maybe `app/config.ts` 🤔

**Comment by VicGUTT** [comment]
Came across this discussion while searching the web for "Nuxt 3 customize directory structure".

I have an upcoming project where we're currently leaning for a directory structure _(inspired by [Laravel](https://github.com/laravel/laravel))_ similar to:

```
.output
.nuxt
app
    Domain specific stuff _(Models, Repositories, ...)_
assets
config
    app.ts _(name, locale, fallback_locale, asset_url, ...)_
    database.ts
composables
client
    middlewares
content
database
    factories
        UserFactory.ts
    migrations
        2023_04_13_000000_create_users_table.ts
    seeders
        UserSeeder.ts
lang
    en
        home.(json|ts)
        other.(json|ts)
    fr
        home.(json|ts)
        other.(json|ts)
layers
modules
node_modules
plugins
public
server
    middlewares
    routes
tests
utils
views
    components
    layouts
    pages
    app.vue
nuxt.config.ts
router.options.ts
```

The idea, is to mainly separate the views _(what's rendered, `.vue`)_ from the domain specific logics, from the various different parts of the app _(database, configs, localization, etc...)_.

If we look closely, the structure we're going with is pretty similar to the one proposed albeit with a few additions such as "database", "config", ... .
The main differences with our structure and the one proposed are the roles given to the "app" directory, the "client" directory, and the aggregation of all `.vue` files into a "views" directory.

The `app` directory will house all the project's core logics, the `views` directory will house the "rendering" layer of the project while the `client` directory will be homologous to the `server` directory basically housing similar concepts but specific to their environnement.

---

_Some might argue the domain specific logic could be made as composables but I'd argue that while they CAN be used in composables they should be a standalone concept. This has the added benefit of simplifying the tests too._

**Comment by VicGUTT** [comment]
In the meantime, if anyone knows how to customize the directory structure in a Nuxt3 app, I would really appreciate your input 😊✌

**Reply by atinux** [reply]
You can by using the `dir` property in your `nuxt.config.ts`, it's fully typed :)

**Comment by mvdschee** [comment]
Introducing a new folder `/app` would make sense, but I'm not sure I agree with all directories that are put there. What I have started to love with the plugins is the optional `.client` and `.server` suffix to have that fine-grain control, but it's an optional tweak you can make if you are ready to optimize your app. If plugins were to be placed in the `/app` directory, it wouldn't make sense to me that it would still work on the server too.

With that in mind, I would like to keep the following directories at the root of the project `composables, middleware, plugins, and utils` with the optional `.client` and `.server` suffixes to split them when needed.

_I'm aware that there is a `middleware/` and `utils/` already in the server directory, and it would remove them in favor of the optional suffixes_

**Reply by atinux** [reply]
Actually the `app/` directory won't prevent anything about `.client` and `.server`, everything in the `app/` directory is part of your Vue application that is universal (SSR + CSR).

**Reply by mvdschee** [reply]
Yeah, I think my mental model is still based around the server and client separation, but `app/` can be both with Nuxt.

how will the `layers/` folder work? would there be a whole new project in there or just some components?

**Comment by masedri** [comment]
For me the solution is a structure base in features, similar a Next 13. Each feature with its own composable, pages, utils, etc.

**Comment by frasza** [comment]
I love the idea and how well structured it looks.

**Comment by Hebilicious** [comment]
Thank you @Atinux, for taking the time to write this RFC.

I really like the fact that we now have `server/routes`, and `plugin` and `modules` are now clearly separated. 

I think that if we are changing the structure, we should consider more changes and bring in more benefits.

- `app` is too generic of a name and doesn't indicates what's inside
- There's now an additional level of nesting for not too much to gain
- `app/router.options.ts` feels out of place
- content is not a part of nuxt/core, why have it here ?
- difference between `utils` and `composables` ?
- why not get rid of both utils and use `composables` everywhere?
- what about `app.config.ts` ? Is it deprecated ?
- why `main.vue` and not `index.vue` ?
- where would `tests` go ? That should be part of the directory structure imo
 
 Instead of `app` and `server`, how about `nuxt` and `nitro` ? This makes things clearer imo.
 Here's what I'm thinking : 
 
 ```bash
 .output/
.build/
node_modules/
nuxt/
  assets/
  components/
  content/ #this should be next to pages
  composables/
  layouts/
  middleware/ 
  pages/
  plugins/
  tests/ # component tests
  index.vue
nitro/
  public/ # this is more consistent with how nitro works
  api/
  middleware
  routes/
  tests/ # server only tests can go here.
  composables/ # better convention with `useX` being  auto-imported
config/ #could be settings
  layers/
  modules/
  nuxt.config.ts # if it's nested in config we could renamed to `nuxt.ts`
  app.config.ts  # could be moved to runtime.config.public
  router.config.ts
tests/ #global/integration/e2e tests can go here
nuxt.config.ts # we could support this as well
```

Minimal starter
```bash
nuxt/
  index.vue
nitro/
  public/
    favicon.ico/
config/
	nuxt.config.ts
```

**Comment by command-tab** [comment]
Jumping in here to mention a scenario to consider: In my web apps' git repos, I usually have a _non-Javascript_ API that's completely separate from my Nuxt app (other than the agreed-upon REST communication), so my repos' directory structure frequently looks something like:

```
api/  # A Python REST API
    Pipfile
    Pipfile.lock
    models/
    resources/
    __init__.py
    etc.
app/  # Nuxt 3 app
    app.vue
    pages/
    composables/
    components/
    etc.
tf/  # Terraform configuration files
    main.tf
    terraform.tfvars
    etc.
```

Moving towards a model where Nuxt has its own `app/` directory would, at least in my case, result in an `app/app/` path, which wouldn't be the end of the world, but it would look a little odd.

This is all to say that it might be worth considering that a given Nuxt app might already live in a directory called `app` to separate it from other services within a single repo, and that not all APIs used by a Nuxt app are ones served by Nuxt/Nitro. It's also possible that this is a highly unusual use case and should be completely ignored, in which case, feel free :) Either way, thank you for continually thinking about ways to improve Nuxt! It truly makes me more productive and my job more enjoyable.

**Comment by ennioVisco** [comment]
> Nuxt 3 is now a fullstack framework thanks to [Nitro](https://nitro.unjs.io) with the ability to have [auto-imported modules](https://nuxt.com/docs/guide/directory-structure/modules) and [layers](https://nuxt.com/docs/getting-started/layers), I believe it's time to rethink the directory structure to give more separation for the "app" level.
> 
> This is the proposal I am thinking of:
> 
> ```shell
> .output/
> .nuxt/
> app/
>   assets/
>   components/
>   composables/
>   layouts/
>   middleware/
>   pages/
>   plugins/
>   utils/
>   main.vue
>   router.options.ts
> content/
> layers/
> modules/
> node_modules/
> public/
> server/
>   api/
>   middleware
>   routes/
>   utils/
> nuxt.config.ts
> ```
> 
> Do you see any downside of this approach?
> 
> The minimal starter would then be:
> 
> ```shell
> app/
>   main.vue
> public/
>   favicon.ico
> nuxt.config.ts
> ```

I jump in to bring a case I have against the word `middleware`.
In fact, I consider the usage of the word as an anti-pattern, because of how vague and misleading it can be.
Looking at Wikipedia, we can find the following definition:

> Middleware is a type of [computer software](https://en.wikipedia.org/wiki/Computer_software) programme that provides services to software applications beyond those available from the [operating system](https://en.wikipedia.org/wiki/Operating_system). It can be described as "software glue"

This definition resonates with how people commonly refer to middleware (see e.g. [Best Middleware Software 2023 - Reviews on 12+ Tools](https://www.getapp.com/it-management-software/middleware/).
I believe an easy analogy is to consider the middleware as the _motherboard_ of the system we are building, to which we attach some higher-level components (e.g. the memory and CPU), as well as the foundational ones (the power supply and the case).

Back to Nuxt, this is not what we mean by middleware: in Nuxt's world, middleware are essentially route guards/adapters/filters: they receive routes, do some processing, and return routes. 
I'd be more in favor in future versions of nuxt, for a more specific word (I really like **guard** to make explicit the kind of behaviours one usually defines with them), and leave the middleware word to the mists of memory 😄

**Reply by typed-sigterm** [reply]
Perhaps `guards` is better?

**Comment by martinszeltins** [comment]
I prefer having my layers be at the top-level because it helps give a better overview of the app and also helps with DDD. I think for large-scale projects it is better to organize the directory structure by **domains** than by **types** Now whenever someone looks at my app he can clearly see what the app is about. So I wouldn't put these layers inside layers/

````
users
  └─ components
  └─ composables
  └─ nuxt.config.ts
settings
  └─ components
  └─ composables
  └─ nuxt.config.ts
sales
  └─ components
  └─ composables
  └─ nuxt.config.ts
app.vue
package.json
nuxt.config.ts
````

**Reply by typed-sigterm** [reply]
Agree, but this might make it difficult to find the source file of a page.

**Comment by typed-sigterm** [comment]
BTW should `middleware` dir be `middlewares`？

The names of other directories are all in plural form, only `middleware/` is in singular form.

**Comment by typed-sigterm** [comment]
`/app` dir is currently using by [`spaLoadingTemplate`](https://nuxt.com/docs/api/nuxt-config#spaloadingtemplate), would it be moved somewhere to avoid conflict with this issue?

**Reply by TheAlexLichter** [reply]
it'd live top-level, like `router.options.ts`

**Comment by bernhardberger** [comment]
Do layers have to follow the same directory structure as main projects to reap the benefits?

```
layers/users
  └─ components
  └─ composables
  └─ nuxt.config.ts
layers/settings
  └─ components
  └─ composables
  └─ nuxt.config.ts
layers/sales
  └─ components
  └─ composables
  └─ nuxt.config.ts
app/app.vue
package.json
nuxt.config.ts
```

vs.

```
layers/users
  └─ app/components
  └─ app/composables
  └─ nuxt.config.ts
layers/settings
  └─ app/components
  └─ app/composables
  └─ nuxt.config.ts
layers/sales
  └─ app/components
  └─ app/composables
  └─ nuxt.config.ts
app/app.vue
package.json
nuxt.config.ts
```

Will a layer not following this structure destroy the performance gains and will nuxt4 fall back to "old structure" mode when a layer doesn't follow the pattern?

This is currently **very** unclear imo.

**Reply by martinszeltins** [reply]
@bernhardberger maybe this will help, it explains how to work with layers in nuxt 4 - https://youtu.be/XGcJiG0fZ8Y

Basically, you want to put things under `/app` directory, the reason being that Nuxt won't have to scan the root directory which includes node_modules etc. That is how you get the performance gains.

**Reply by bernhardberger** [reply]
> @bernhardberger maybe this will help, it explains how to work with layers in nuxt 4 - https://youtu.be/XGcJiG0fZ8Y
> 
> Basically, you want to put things under `/app` directory, the reason being that Nuxt won't have to scan the root directory which includes node_modules etc. That is how you get the performance gains.

This directly contradicts what @manniL is suggesting in his Nuxt4 Directory Structure video where he puts the `layers` folder into `rootDir` (which makes sense to me as layers can have `server` parts which should not be under /app).

**Reply by martinszeltins** [reply]
@bernhardberger I see what you mean. There are multiple ways to work with layers. The `layers` directory is a special directory that auto registers the layers for you. So it goes at the root level. It is also perfectly fine to put them under /app/layerX and in that case you can also put server under app/layerX/server. Multiple ways to do the same thing.

**Reply by bernhardberger** [reply]
But what are the implications of either of those scenarios?

What are the benefits of the srcDir (app) when we can just randomly stuff directories into the rootDir (like types and interfaces that are shared between server and app), place layers with server parts into `/app`, etc?

Wouldn't nuxt "auto detect" "old structure" and fall back into the mode with all it's negative impact when I put random directories like `types` into the rootDir?

Wouldn't all benefits be gone too as soon as I put a /server dir into a layer?

Why would I get all the benefits when I don't  use an srcDir (app) folder *inside* a layer (which effectively gets copied and merged at build time..?)? Isn't also the argument that both /app and /server run in different TS contexts? How does my IDE/TypeScript lang service distinguish that then? Doesn't it have the same problems as before that the /app folder is trying to solve?

**Reply by yosefeliezrie** [reply]
> @bernhardberger I see what you mean. There are multiple ways to work with layers. The `layers` directory is a special directory that auto registers the layers for you. So it goes at the root level. It is also perfectly fine to put them under /app/layerX and in that case you can also put server under app/layerX/server. Multiple ways to do the same thing.

I know someone may have already recommendes but I would strongly recommend you watch the video from nuxt core member @manniL after quite a bit of testing it seems he's method is best and gets the best performance and also for me makes the most sense. 

https://youtu.be/KnCNOp5Pbfs?si=fnycOw89WN8fq3bM

I would watch the full Video as well

https://youtu.be/KnCNOp5Pbfs

~~~
layers/users
  └─ components
  └─ composables
  └─ nuxt.config.ts
layers/settings
  └─ components
  └─ composables
  └─ nuxt.config.ts
layers/sales
  └─ components
  └─ composables
  └─ nuxt.config.ts
app/app.vue
package.json
nuxt.config.ts
~~~

**Comment by RasmusPedersenKnowit** [comment]
@atinux @pi0 - Now that this is the default for new projects. Is the fallback/scan for the old structure, going to be deprecate the scan for the old structure in future major releases ?
